STEP/ArduinoCompiler.cs

```diff
@@ -16,40 +19,148 @@ public class ArduinoCompiler
+                    $"{directoryPath}/compiled.c -o {directoryPath}/compiled.out");
+
+                compiler?.WaitForExit();
+
+                if (!File.Exists(directoryPath + "/compiled.out"))
+                {
+                    throw new ApplicationException("Arduino compiler error. compiled.out file was not generated.");
+                }
+
+                Process hexConverter = Process.Start("/bin/bash",
+                    $"avr-objcopy -O ihex {directoryPath}/compiled.out {directoryPath}/compiled.hex");
+
+                hexConverter?.WaitForExit();
+            }
+
+            if (!File.Exists(directoryPath + "/compiled.hex"))
+            {
+                throw new ApplicationException("Arduino compiler error. compiled.hex file was not generated.");
+            }
+
+            // Enable upload logging
+            var nlogConfig = new LoggingConfiguration();
+
+            nlogConfig.AddRule(minLevel: LogLevel.Trace, maxLevel: LogLevel.Fatal,
+                target: new ConsoleTarget("consoleTarget")
+                {
+                    Layout = "${longdate} level=${level} message=${message}"
+                });
+
+            LogManager.Configuration = nlogConfig;
+
+            // Update compiled hex file to Arduino board
+            ArduinoSketchUploader uploader = new ArduinoSketchUploader(new ArduinoSketchUploaderOptions()
+            {
+                FileName = directoryPath + "/compiled.hex",
+                PortName = "COM3", // Can be omitted to try to auto-detect the COM port.
+                ArduinoModel = ArduinoModel.UnoR3
+            }, new NLogArduinoUploaderLogger(error: true, warn: true, info: true));
+            uploader.UploadSketch();
+        }
+        catch (ApplicationException e)
         {
-            // The /C flag means that cmd should execute the following command and exit without waiting for further input.
-            Process.Start("cmd.exe",
-                "/C " +
-                $"avr-gcc -O2 -Wall -mmcu=atmega328p {directoryPath}/compiled.c -o {directoryPath}/compiled.out");
-            Process.Start("cmd.exe",
-                "/C " + $"avr-objcopy -O ihex {directoryPath}/compiled.out {directoryPath}/.compiled.hex");
+            Console.WriteLine(e);
         }
-        else
+        finally
         {
-            Process.Start("/bin/bash",
-                $"avr-gcc -O2 -Wall -mmcu=atmega328p {directoryPath}/compiled.c -o {directoryPath}/compiled.out");
-            Process.Start("/bin/bash",
-                $"avr-objcopy -O ihex {directoryPath}/compiled.out {directoryPath}/.compiled.hex");
+            // Clean up
+            if (File.Exists(directoryPath + "/compiled.out"))
+            {
+                File.Delete(directoryPath + "/compiled.out");
+            }
+
+            if (File.Exists(directoryPath + "/compiled.hex"))
+            {
+                File.Delete(directoryPath + "/compiled.hex");
+            }
```

Comment:
God detalje!

Reply:
Og apropros, så kan man måske sørge for at den kompilerede fil har samme navn som input-filen?

Reply:
Filnavne følger nu med igennem programmet.